pkg/collector/runner/num_workers_test.go

```diff
@@ -0,0 +1,227 @@
+package runner
+
+import (
+	"runtime"
+	"strconv"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	"github.com/DataDog/datadog-agent/pkg/collector/py"
+	python "github.com/sbinet/go-python"
+)
+
+/****************** Testing configuration ****************************/
+
+var testingEfficiency = false // Run this test (should be false normally)
+var static = false            // Run with default num workers (vs dynamic)
+var prettyOutput = false      // Labelled test results
+var checkType = busyWait      // Which type of check to run (busyWait, lazyWait, or pythonCheck)
+var numIntervals = 10         // How many times to repeat each check (for the time tests)
+var memoryTest = false        // After the time tests, run the memory test
+
+/*********************************************************************/
+
+type CheckType int
+
+const (
+	busyWait    CheckType = iota
+	lazyWait    CheckType = iota
+	pythonCheck CheckType = iota // BUG: testing the python checks crashes every so often
+)
+
+type stickyLock struct {
+	gstate python.PyGILState
+	locked uint32 // Flag set to 1 if the lock is locked, 0 otherwise
+}
+
+// NumWorkersCheck implements the 'Check' interface via the TestCheck struct defined in runner_test.go
+type NumWorkersCheck struct {
+	TestCheck
+	name string
+}
+
+func (nc *NumWorkersCheck) String() string { return nc.name }
+func (nc *NumWorkersCheck) ID() check.ID   { return check.ID(nc.String()) }
+
+func (nc *NumWorkersCheck) Run() error {
+	switch checkType {
+	case busyWait:
+		start := time.Now()
+		now := time.Now()
+		for {
+			if now.Sub(start) > time.Millisecond*100 {
+				break
+			}
+			now = time.Now()
+		}
+
+	case lazyWait:
+		time.Sleep(time.Millisecond * 100)
+
+	case pythonCheck:
+		runPythonCheck()
+	}
+
+	nc.hasRun = true
+	return nil
+}
+
+// Efficiency test for the number of check workers running
+// Note: use the -v flag when testing to see the output
+func TestUpdateNumWorkers(t *testing.T) {
+	if !testingEfficiency {
+		return
+	}
+
+	t.Logf("Testing efficiency on a machine with %v cores", runtime.NumCPU())
+	if checkType == pythonCheck {
+		// Initialize the python interpreter
+		state := py.Initialize(".", "../dist")
+		defer python.PyEval_RestoreThread(state)
+	}
+
+	// Run the time tests
+	interval := false
+	for i := 0; i < 2; i++ {
+		if interval {
+			t.Logf("********* Starting the time efficiency test (%v repeats) *********", numIntervals)
+		} else {
+			t.Log("********* Starting the time efficiency test (single) *********")
+		}
+
+		checksToRun := [10]int{5, 15, 25, 35, 45, 55, 65, 75, 85, 100}
+
+		for _, n := range checksToRun {
+			ti := timeToComplete(n, interval)
+
+			if prettyOutput {
+				t.Logf("Time to run %v checks: %v", n, ti.Seconds())
+			} else {
+				t.Logf("%v", ti.Seconds())
+			}
+		}
+
+		if numIntervals == 0 {
+			break
+		}
+		interval = true
+	}
+
+	if !memoryTest {
+		return
+	}
+	// Run the memory test
```

Comment:
I think the use case for this particular test would be as a standalone test, not running the entire test suite for the runner, which requires passing the name of the function you want to run as an argument to `go test`... so I think it would be a little awkward to have to pass 2 separate test names to pass... I can pull it into a separate function to clean up the code a little but I think it's still better left as a flag on this one test function. Thoughts?